File: web/src/lib/glow.ts
import type { ScoutItem } from './wasm/types'

/**
 * Enchantment / glyph glow colours and pulse periods, mirrored 1:1 from
 * Shattered Pixel Dungeon's `ItemSprite.Glowing` definitions so the scout icons
 * pulse exactly as the game renders them. `period` is the seconds the glow takes
 * to fade fully in — it fades back out over the same span, so a complete pulse
 * cycle lasts `2 × period`. Curses always glow black, matching the game.
 */
export interface Glow {
  /** Hex colour the sprite blends toward at the pulse peak. */
  color: string
  /** Seconds to reach peak glow; the full fade-in/out cycle lasts twice this. */
  period: number
}

/** Upstream's default `Glowing(color)` period when none is given (1f). */
const DEFAULT_PERIOD = 1

// Keyed by the wire names the scout emits (WeaponEffect / ArmorEffect
// `wire_name` in seedfinder-core). Only non-curse effects live here; every curse
// glows black and is handled by CURSE_GLOW below.
const ENCHANT_GLOW: Record<string, Glow> = {
  // Weapon enchantments
  Blazing: { color: '#ff4400', period: DEFAULT_PERIOD },
  Chilling: { color: '#00ffff', period: DEFAULT_PERIOD },
  Kinetic: { color: '#ffff00', period: DEFAULT_PERIOD },
  Shocking: { color: '#ffffff', period: 0.5 },
  Blocking: { color: '#0000ff', period: DEFAULT_PERIOD },
  Blooming: { color: '#008800', period: DEFAULT_PERIOD },
  Elastic: { color: '#ff00ff', period: DEFAULT_PERIOD },
  Lucky: { color: '#00ff00', period: DEFAULT_PERIOD },
  Projecting: { color: '#8844cc', period: DEFAULT_PERIOD },
  Unstable: { color: '#999999', period: DEFAULT_PERIOD },
  Corrupting: { color: '#440066', period: DEFAULT_PERIOD },
  Grim: { color: '#000000', period: DEFAULT_PERIOD },
  Vampiric: { color: '#660022', period: DEFAULT_PERIOD },
  // Armor glyphs
  Obfuscation: { color: '#888888', period: DEFAULT_PERIOD },
  Swiftness: { color: '#ffff00', period: DEFAULT_PERIOD },
  Viscosity: { color: '#8844cc', period: DEFAULT_PERIOD },
  Potential: { color: '#ffffff', period: 0.6 },
  Brimstone: { color: '#ff4400', period: DEFAULT_PERIOD },
  Stone: { color: '#222222', period: DEFAULT_PERIOD },
  Entanglement: { color: '#663300', period: DEFAULT_PERIOD },
  Repulsion: { color: '#ffffff', period: DEFAULT_PERIOD },
  Camouflage: { color: '#448822', period: DEFAULT_PERIOD },
  Flow: { color: '#0000ff', period: DEFAULT_PERIOD },
  Affection: { color: '#ff4488', period: DEFAULT_PERIOD },
  'Anti-Magic': { color: '#88eeff', period: DEFAULT_PERIOD },
  Thorns: { color: '#660022', period: DEFAULT_PERIOD },
}

/** Every curse glows black in the game, at the default period. */
const CURSE_GLOW: Glow = { color: '#000000', period: DEFAULT_PERIOD }

/**
 * The pulse glow for a scouted item, or null when it carries no enchantment or
 * curse. A beneficial enchantment/glyph wins even on a cursed item (matching
 * `Weapon.glowing()`, which returns the enchantment's colour when one is
 * present — e.g. a curse-infused Kinetic weapon still glows yellow); otherwise a
 * cursed item pulses black.
 */
export function itemGlow(item: Pick<ScoutItem, 'cursed' | 'effect'>): Glow | null {
  if (item.effect?.kind === 'enchantment') return ENCHANT_GLOW[item.effect.name] ?? null
  if (item.cursed) return CURSE_GLOW
  return null
}

/**
 * The pulse glow for a bare effect name (as carried by a requirement), or null
 * when there is none. Known enchantments/glyphs pulse their colour; any other
 * effect name is a curse and pulses black.
 */
export function effectGlow(effect: string | undefined): Glow | null {
  if (!effect) return null
  return ENCHANT_GLOW[effect] ?? CURSE_GLOW
}

File: android/app/src/main/java/dev/seedseeker/app/ui/Glow.kt
// SPDX-License-Identifier: GPL-3.0-or-later
package dev.seedseeker.app.ui

import android.content.Context
import android.provider.Settings
import androidx.compose.animation.core.withInfiniteAnimationFrameMillis
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.LongState
import androidx.compose.runtime.Stable
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import dev.seedseeker.app.catalog.ItemCatalog
import dev.seedseeker.app.model.EffectFilter
import dev.seedseeker.app.model.ItemKind
import kotlin.math.abs

/**
 * An enchantment / curse glow: the colour the sprite blends toward at the pulse
 * peak, plus the seconds it takes to reach that peak. The glow fades back out
 * over the same span, so one full cycle lasts `2 × period`.
 */
data class Glow(val color: Color, val period: Float)

/**
 * Enchantment / glyph glow colours and pulse periods, mirrored 1:1 from
 * Shattered Pixel Dungeon's `ItemSprite.Glowing` definitions — and kept in step
 * with the web front-end's `web/src/shared/sprites/glow.ts`, which is keyed by the same
 * wire names the scout emits. Curses are absent from the table because weapon and armor
 * curse effects glow black in the game; the catalog names them.
 */
object ItemGlows {
    /** Upstream's default `Glowing(color)` period when none is given (1f). */
    private const val DEFAULT_PERIOD = 1f

    private val enchantments: Map<String, Glow> = mapOf(
        // Weapon enchantments
        "Blazing" to Glow(Color(0xFFFF4400), DEFAULT_PERIOD),
        "Chilling" to Glow(Color(0xFF00FFFF), DEFAULT_PERIOD),
        "Kinetic" to Glow(Color(0xFFFFFF00), DEFAULT_PERIOD),
        "Shocking" to Glow(Color(0xFFFFFFFF), 0.5f),
        "Venomous" to Glow(Color(0xFF4400AA), DEFAULT_PERIOD),
        "Blocking" to Glow(Color(0xFF0000FF), DEFAULT_PERIOD),
        "Blooming" to Glow(Color(0xFF008800), DEFAULT_PERIOD),
        "Eldritch" to Glow(Color(0xFF222222), DEFAULT_PERIOD),
        "Elastic" to Glow(Color(0xFFFF00FF), DEFAULT_PERIOD),
        "Lucky" to Glow(Color(0xFF00FF00), DEFAULT_PERIOD),
        "Projecting" to Glow(Color(0xFF8844CC), DEFAULT_PERIOD),
        "Unstable" to Glow(Color(0xFF999999), DEFAULT_PERIOD),
        "Vorpal" to Glow(Color(0xFFAA6666), DEFAULT_PERIOD),
        "Corrupting" to Glow(Color(0xFF440066), DEFAULT_PERIOD),
        "Crystal" to Glow(Color(0xFF0088FF), DEFAULT_PERIOD),
        "Grim" to Glow(Color(0xFF000000), DEFAULT_PERIOD),
        "Vampiric" to Glow(Color(0xFF660022), DEFAULT_PERIOD),
        // Armor glyphs
        "Obfuscation" to Glow(Color(0xFF888888), DEFAULT_PERIOD),
        "Swiftness" to Glow(Color(0xFFFFFF00), DEFAULT_PERIOD),
        "Viscosity" to Glow(Color(0xFF8844CC), DEFAULT_PERIOD),
        "Potential" to Glow(Color(0xFFFFFFFF), 0.6f),
        "Brimstone" to Glow(Color(0xFFFF4400), DEFAULT_PERIOD),
        "Stone" to Glow(Color(0xFF222222), DEFAULT_PERIOD),
        "Entanglement" to Glow(Color(0xFF663300), DEFAULT_PERIOD),
        "Repulsion" to Glow(Color(0xFFFFFFFF), DEFAULT_PERIOD),
        "Camouflage" to Glow(Color(0xFF448822), DEFAULT_PERIOD),
        "Flow" to Glow(Color(0xFF0000FF), DEFAULT_PERIOD),
        "Affection" to Glow(Color(0xFFFF4488), DEFAULT_PERIOD),
        "Anti-Magic" to Glow(Color(0xFF88EEFF), DEFAULT_PERIOD),
        "Thorns" to Glow(Color(0xFF660022), DEFAULT_PERIOD),
    )

    /** Curse glow for equipment other than wands, at the default period. */
    private val curse = Glow(Color(0xFF000000), DEFAULT_PERIOD)

    /** The catalog's own curse names, for weapons and armor alike. */
    private val curses: Set<String> by lazy {
        (ItemCatalog.cursesFor(ItemKind.WEAPON) + ItemCatalog.cursesFor(ItemKind.ARMOR)).toSet()
    }

    /**
     * The pulse glow for a scouted item, or null when it carries no enchantment
     * or curse. A beneficial enchantment/glyph wins even on a cursed item
     * (matching `Weapon.glowing()`, which returns the enchantment's colour when
     * one is present — a curse-infused Kinetic weapon still glows yellow);
     * otherwise cursed equipment other than wands pulses black. Wands never glow.
     */
    fun forItem(kind: ItemKind, effect: String?, cursed: Boolean): Glow? =
        if (kind == ItemKind.WAND) null
        else effect?.let(::forEffect) ?: if (cursed) curse else null

    /**
     * The pulse glow for a bare effect name (as carried by a requirement), or
     * null when there is none. Enchantments and glyphs pulse their own colour;
     * the catalog's curses pulse black.
     */
    fun forEffect(effect: String?): Glow? {
        if (effect == null) return null
        return enchantments[effect] ?: curse.takeIf { effect in curses }
    }

    /**
     * The glows a requirement's effect filter pulses through, in the filter's
     * own order — one per named effect. "Any effect" and "any enchantment"
     * settle on no colour of their own and so glow with none.
     */
    fun forFilter(filter: EffectFilter): List<Glow> = when (filter) {
        EffectFilter.Any, EffectFilter.AnyEnchantment -> emptyList()
        is EffectFilter.OneOf -> filter.names.map { enchantments[it] ?: curse }
    }

}

/** Peak blend toward the glow colour, matching the web's `d1-ench-pulse`. */
private const val GLOW_PEAK_ALPHA = 0.6f

/** Frozen blend used when the system asks for reduced motion. */
internal const val GLOW_STATIC_ALPHA = 0.3f

/** The colour a sprite is tinted with right now, and how far toward it. */
data class GlowBlend(val color: Color, val alpha: Float)

/**
 * One shared pulse clock for every glowing sprite on screen. Sprites read the
 * elapsed time inside their draw lambda, so a frame invalidates drawing only —
 * never composition — and dozens of scout rows animate off a single frame
 * subscription instead of one infinite transition each.
 */
@Stable
class GlowPulse internal constructor(
    private val elapsedMillis: LongState,
    private val animated: Boolean,
) {
    /**
     * The blend factor for a glow of this period: linear 0 → 0.6 → 0 across
     * `2 × period` seconds, reproducing upstream's `texel*(1-v) + glow*v` shader
     * exactly as the web animates it. Held at a static 0.3 under reduced motion.
     */
    fun alphaFor(period: Float): Float {
        if (!animated) return GLOW_STATIC_ALPHA
        val cycleMillis = cycleOf(period)
        val phase = (elapsedMillis.longValue % cycleMillis) / cycleMillis.toFloat()
        return GLOW_PEAK_ALPHA * (1f - abs(2f * phase - 1f))
    }

    /**
     * What a stack of glows shows now. Each takes the sprite for one whole
     * cycle of its own before handing it on, so an item asking for three
     * effects pulses all three in a round, and the round lasts the sum of
     * their cycles — the same round the board's effect badge wipes through.
     * Frozen on the first glow under reduced motion, where a hand-over the
     * user cannot see would only leave one colour picked arbitrarily.
     */
    fun blendFor(glows: List<Glow>): GlowBlend? {
        val first = glows.firstOrNull() ?: return null
        if (!animated) return GlowBlend(first.color, GLOW_STATIC_ALPHA)
        var position = elapsedMillis.longValue % glows.sumOf { cycleOf(it.period) }
        for (glow in glows) {
            val cycle = cycleOf(glow.period)
            if (position < cycle) {
                val phase = position / cycle.toFloat()
                return GlowBlend(glow.color, GLOW_PEAK_ALPHA * (1f - abs(2f * phase - 1f)))
            }
            position -= cycle
        }
        return GlowBlend(first.color, 0f)
    }
}

/** One full pulse of a glow: up to the peak and back, in milliseconds. */
private fun cycleOf(period: Float): Long = (2_000f * period).toLong().coerceAtLeast(1L)

private val StaticGlowPulse = GlowPulse(mutableLongStateOf(0L), animated = false)

/** The pulse clock in scope; a frozen one when no host provided a running clock. */
val LocalGlowPulse = staticCompositionLocalOf { StaticGlowPulse }

/** Starts the app-wide glow clock, or a frozen one when animations are off. */
@Composable
fun rememberGlowPulse(): GlowPulse {
    val context = LocalContext.current
    val animated = remember(context) { animationsEnabled(context) }
    val elapsed = remember { mutableLongStateOf(0L) }
    LaunchedEffect(animated) {
        if (!animated) return@LaunchedEffect
        var origin = -1L
        while (true) {
            withInfiniteAnimationFrameMillis { frameMillis ->
                if (origin < 0L) origin = frameMillis
                elapsed.longValue = frameMillis - origin
            }
        }
    }
    return remember(elapsed, animated) { GlowPulse(elapsed, animated) }
}

/**
 * False when the user has turned animations off — the developer-options animator
 * scale and the "remove animations" accessibility toggle both zero this setting.
 */
internal fun animationsEnabled(context: Context): Boolean =
    Settings.Global.getFloat(
        context.contentResolver,
        Settings.Global.ANIMATOR_DURATION_SCALE,
        1f,
    ) != 0f
